### ISSUE TITLE
DownloadObjectAsync hangs

### ISSUE THREAD

**Issue by atdimitrov** [googleapis/google-cloud-dotnet#8225]
Hello. We are experiencing a problem when trying to download portion of a file with `DownloadObjectAsync`. The code is pretty simple:

```csharp
await client.DownloadObjectAsync(bucketName, objectKey, outputStream, new DownloadObjectOptions()
{
    Range = new RangeHeaderValue(offset, length)
});
```

The setup however is a bit more complex: we have a distributed system of hundreds of applications running the code above, all of them downloading different parts of the same file (around 400GB in size) hosted in GCP. The problem is that some of those calls to `DownloadObjectAsync` hang and do not return (it is hitting 20 min timeout in different part of the system; normally 1 such call takes at most 4-5 mins, with average of under 1 minute).

Replacing the implementation with [this one](https://cloud.google.com/storage/docs/downloading-objects#downloading-an-object-portion) fixes the problem, but this solution includes implementation details to our codebase that we would like to abstract using the SDK.

Any ideas?

#### Environment details

  - Base Docker image: mcr.microsoft.com/azure-functions/dotnet:3.0
  - OS: Debian 10
  - .NET version: .NET Core 3.1
  - Package name and version: Google.Cloud.Storage.V1 3.7.0

#### Steps to reproduce
Couldn't reproduce this in an isolated environment.

**Comment by jskeet** [comment]
It's going to be quite hard to help without being able to reproduce this, unfortunately. A few questions that might shed some light:

- You mention the Azure Functions base docker image - is this actually running in Azure Functions, or are you just using that as the docker image?
- Is it possible that there's an asynchronous context issue? Have you tried using `.ConfigureAwait(false)` on the task?
- Have you tried passing an `IProgress<IDownloadProgress>` to log each time progress is made? That would give more of an idea of whether it's being really slow but making progress, or just stopping entirely, or perhaps not even getting a single chunk
- Are you able to try passing a cancellation token in with a timeout (e.g. `new CancellationTokenSource(TimeSpan.FromMinutes(10)).Token`)? That might give some indication as to where things are hanging

**Comment by atdimitrov** [comment]
> You mention the Azure Functions base docker image - is this actually running in Azure Functions, or are you just using that as the docker image?

Yes, it is running as an Azure Function in Kubernetes. Do you think it is related in any way to the issue?

> Is it possible that there's an asynchronous context issue? Have you tried using .ConfigureAwait(false) on the task?

I believe it is not, but there's a chance it is. We can easily try with `ConfigureAwait`. We could also try to use the synchronous API of the SDK but the may be in the asynchronous version of the API only.

> Have you tried passing an IProgress<IDownloadProgress> to log each time progress is made? That would give more of an idea of whether it's being really slow but making progress, or just stopping entirely, or perhaps not even getting a single chunk

Will try that.

> Are you able to try passing a cancellation token in with a timeout (e.g. new CancellationTokenSource(TimeSpan.FromMinutes(10)).Token)? That might give some indication as to where things are hanging

Yes, we can do that.

> What is the output stream you're writing to? Is it possible that it's the write to that that is hanging?

It is a `MemoryStream`. I don't expect this to be able to hang. If anything we should get `OutOfMemoryException`, but we haven't.

I will perform those tests and I'll get back to you with the results.

**Comment by amanda-tarafa** [comment]
Also, this might be realted to [Bandwidth quota](https://cloud.google.com/storage/quotas#bandwidth). Note that some requests may be throtled when quota is exceeded.

> When a project's bandwidth exceeds quota in a location, requests to affected buckets can be rejected with a [retryable](https://cloud.google.com/storage/docs/retry-strategy) 429 error or can be throttled.

See here how you can monitor your [Bandwith Usage](https://cloud.google.com/storage/docs/bandwidth-usage) to check if this is related.

**Comment by atdimitrov** [comment]
> Also, this might be realted to [Bandwidth quota](https://cloud.google.com/storage/quotas#bandwidth). Note that some requests may be throtled when quota is exceeded.
> 
> > When a project's bandwidth exceeds quota in a location, requests to affected buckets can be rejected with a [retryable](https://cloud.google.com/storage/docs/retry-strategy) 429 error or can be throttled.
> 
> See here how you can monitor your [Bandwith Usage](https://cloud.google.com/storage/docs/bandwidth-usage) to check if this is related.

Yes, throttling was the first thing I suspected as something that may be causing this, but it is very inconsistent - e.g. a failing request (timeouting at the 20 min mark) may be started at roughly the same time as other requests that took 5-10 seconds to finish. Moreover, the other approach with the raw HTTP client didn't see any of that.

To give more context - every request is trying to get ~100MB of data. Is throttling supposed to slow such a request that dramatically (100MB cannot be downloaded for 20 minutes)? Note that the network speed should not be a problem in that case, since it is running in a data center (and also I've seen requests for 100MB finish in under a second).

**Comment by amanda-tarafa** [comment]
> Is throttling supposed to slow such a request that dramatically (100MB cannot be downloaded for 20 minutes)?

That very much depends on the throtling strategy, which I'm not familiar with. But you did say you had hundreds of aplications running at the same time, and that is definetely in the right order of magnitude to at least blow the per region egress quota of 50Gbs. But it's easy to check whether this might be a factor or not, just follow the instructions on [Bandwith Usage](https://cloud.google.com/storage/docs/bandwidth-usage) and see what the quoata graphs are showing. If you are reaching quota then that's, probably, at least part of the problem.

**Comment by atdimitrov** [comment]
> Is it possible that there's an asynchronous context issue? Have you tried using .ConfigureAwait(false) on the task?

So, allowing for continuation on a different synchronization context using `ConfigureAwait` did not fix the issue.

> Have you tried passing an IProgress<IDownloadProgress> to log each time progress is made? That would give more of an idea of whether it's being really slow but making progress, or just stopping entirely, or perhaps not even getting a single chunk

The test was performed with the 10 minute timeout cancellation token. We had 12 timeouts where:
- 2 instances reported up to 80 MB downloaded
- 1 instance reported up to 70 MB downloaded
- 1 instance reported up to 50 MB downloaded
- 3 instances reported up to 40 MB downloaded
- 2 instances reported up to 10 MB downloaded
- 3 instances reported only 0 MB downloaded

Interesting fact is that on a second attempt, we had 19 more timeouts, but it didn't result in `OperationCanceledException` - it just continued working as normal, i.e. the method returned successfully.

Does some of that shed some light on what is it actually going on?

**Comment by jskeet** [comment]
> Interesting fact is that on a second attempt, we had 19 more timeouts, but it didn't result in OperationCanceledException - it just continued working as normal, i.e. the method returned successfully.

I don't understand what you mean here. How did you observe the timeouts, if the method returned successfully?

With the instances that did manage to download some data, were you able to observe whether they were "trickling" data all the time (just not quickly enough to complete before the end), or did they manage to download quite a lot quickly, and then stop entirely?

**Comment by atdimitrov** [comment]
> I don't understand what you mean here. How did you observe the timeouts, if the method returned successfully?

Maybe should've elaborated a bit more. We're measuring the time it takes for `DownloadObjectAsync` to finish. Using that I can see that those 19 calls took a little bit over 10 minutes to return (talking seconds over 10 minutes). Normally the times are much lower (under the 4 minute mark), so it must be because of the timeout.

>> With the instances that did manage to download some data, were you able to observe whether they were "trickling" data all the time (just not quickly enough to complete before the end), or did they manage to download quite a lot quickly, and then stop entirely?

Unfortunately, no. Haven't setup any network monitoring yet, nor I saved the timestamps of the progress notifications (and I should've). Do you think it can be just very severe throttling? If true, we are talking speeds under 17 KB/s for those 3 instances where that reported no higher than 0 MB.

**Comment by jskeet** [comment]
I'm not familiar with the potential severity of the throttling - but I think it's at least plausible, yes.

One thing you might want to try - if you're able to do this on a "scheduled test" basis - is download the same object via `gsutil` from an Azure VM (or potentially just from your development machine) - that shows the download speed.

**Comment by atdimitrov** [comment]
We've worked around the problem with a different approach that has other benefits too, so this is no longer an issue for us. Thank you and @amanda-tarafa for your support.

P.S. Personally, I would be happy to help if you decide to persue the problem.
